cpp/src/arrow/io/interfaces.cc

```diff
@@ -160,6 +163,63 @@ void CloseFromDestructor(FileInterface* file) {
   }
 }
 
+#ifndef NDEBUG
+
+// Debug mode concurrency checking
+
+struct SharedExclusiveChecker::Impl {
+  std::mutex mutex;
+  int64_t n_shared = 0;
```

Comment:
I'd make the following atomics.

Reply:
Why so? There are only read and written with the mutex taken.

Reply:
I spent too much time reading on std::mutex, it is explained in the [Release-Acquire ordering](https://en.cppreference.com/w/cpp/atomic/memory_order) section.